File: Diagrammatize.Design.Pattern/code/python/Interpreter/A1/language/Context.py

from .ParseException import ParseException
from .ExecutorFactory import ExecutorFactory

class Context(ExecutorFactory):
    def __init__(self, text):
        self._tokenizer = text.split()
        self._index = 0
        self._currentToken = None
        self._factory = None
        self.nextToken()
    
    def nextToken(self):
        if self._index < len(self._tokenizer):
            self._currentToken = self._tokenizer[self._index]
            self._index += 1
        else:
            self._currentToken = None
        return self._currentToken
    
    def currentToken(self):
        return self._currentToken
    
    def skipToken(self, token):
        if token != self._currentToken:
            raise ParseException("Warning: " + token + " is expected, but " + str(self._currentToken) + " is found.")
        self.nextToken()
    
    def currentNumber(self):
        if self._currentToken is None:
            raise ParseException("Warning: No token available")
        
        try:
            return int(self._currentToken)
        except ValueError as e:
            raise ParseException("Warning: " + str(e))

    def setExecutorFactory(self, factory):
        self._factory = factory

    def createExecutor(self, name):
        return self._factory.createExecutor(name)

File: Diagrammatize.Design.Patterns/code/cxx/Interpreter/Sample2/Context.cpp
#include "Context.hpp"
#include "ParseException.hpp"

Context::Context(const std::string& text): input_(text) {
    c_ = input_[p_];
    nextToken();
}

void Context::consume() {
    p_++;
    if (p_ >= input_.length()) c_ = EOL;
    else c_ = input_[p_];
}

void Context::WS() {
    while (c_ == ' ' || c_ == '\t' || c_ == '\n' || c_ == '\r')
        consume();
}

std::string Context::nextToken() {
    while (c_ != EOL) {
        switch (c_) {
            case ' ': case '\t': case '\n': case '\r': WS(); continue;
            default:
                if (isLETTER()) return NAME();
                else if (isDIGIT()) return NUMBER();
                else throw ParseException(std::string("invalid charater: ")+c_);
        }
    }

    currentToken_ = "";
    return currentToken_;
}

void Context::skipToken(const std::string& token) {
    if (currentToken_ != token) {
        throw ParseException("Warning: " + token + " is expected, but " + currentToken_ + " is found.");
    }
    nextToken();
}

int Context::currentNumber() {
    int number = 0;
    try {
        number = std::stoi(currentToken_);
    } catch (const std::exception& e) {
        throw ParseException("Warning: " + std::string(e.what()));
    }

    return number;
}

std::string Context::NAME() {
    currentToken_.clear();
    do {
        currentToken_ += c_;
        consume();
    } while (isLETTER());
    return currentToken_;
}

std::string Context::NUMBER() {
    currentToken_.clear();
    do {
        currentToken_ += c_;
        consume();
    } while (isDIGIT());
    return currentToken_;
}
